### PR TITLE
Cm 280 delete files

### PR DESCRIPTION
**Pull request by dcmcand**
**Description of change**
Adds support to delete files.

**How to test**
1. Pull down changes.
2. Upload files.
3. Verify they are present in minio and the db.
4. click on the trash can icon next to the file
5. verify they are gone from minio and the db

**Issue(s)**
* https://github.com/HHS/Head-Start-TTADP/issues/280

**Checklist**

- [X] Meets issue criteria
- [X] Code tested
- [X] Meets accessibility standards (WCAG 2.1 Levels A, AA)
- [X] Documentation updated
    - API methods
    - Boundary and Data Flow Diagrams
    - [Architectural Decision Records](https://adr.github.io/) written for major infrastructure decisions with the [Nygard template](https://github.com/joelparkerhenderson/architecture_decision_record/blob/master/adr_template_by_michael_nygard.md)
    - OSCAL templates completed when security controls are implemented or modified

### CODE REVIEW

**Review by jasalisbury — COMMENTED** [review]
It might be too easy to delete files, should we add a confirmation somewhere so someone doesn't accidentally delete a file? Also do we want to actually delete the file from the database/s3 or just soft delete it (have a `deleted` column that gets set to true in the DB on delete)?

**Comment by dcmcand** [comment]
> It might be too easy to delete files, should we add a confirmation somewhere so someone doesn't accidentally delete a file? Also do we want to actually delete the file from the database/s3 or just soft delete it (have a `deleted` column that gets set to true in the DB on delete)?

I feel like it is good since it isn't like deleting a file on your computer, just removing it from your report. You would still have a copy locally. For example, if you attach a file to an email, then delete it, Gmail doesn't ask for confirmation. However, if the consensus is needing a confirmation box, we can definitely do that.
I had considered the soft delete option, but was having trouble coming up with a use case where that would be helpful. Thoughts?

**Comment by jasalisbury** [comment]
> > It might be too easy to delete files, should we add a confirmation somewhere so someone doesn't accidentally delete a file? Also do we want to actually delete the file from the database/s3 or just soft delete it (have a `deleted` column that gets set to true in the DB on delete)?
> 
> I feel like it is good since it isn't like deleting a file on your computer, just removing it from your report. You would still have a copy locally. For example, if you attach a file to an email, then delete it, Gmail doesn't ask for confirmation. However, if the consensus is needing a confirmation box, we can definitely do that.
> I had considered the soft delete option, but was having trouble coming up with a use case where that would be helpful. Thoughts?

I could see a situation where a collaborator deletes a file. Or a user uploads a file and deletes it locally for some reason (to save space?). Maybe we can tell the users to not delete files locally until the report has been approved?

**Comment by dcmcand** [comment]
> > > It might be too easy to delete files, should we add a confirmation somewhere so someone doesn't accidentally delete a file? Also do we want to actually delete the file from the database/s3 or just soft delete it (have a `deleted` column that gets set to true in the DB on delete)?
> > 
> > 
> > I feel like it is good since it isn't like deleting a file on your computer, just removing it from your report. You would still have a copy locally. For example, if you attach a file to an email, then delete it, Gmail doesn't ask for confirmation. However, if the consensus is needing a confirmation box, we can definitely do that.
> > I had considered the soft delete option, but was having trouble coming up with a use case where that would be helpful. Thoughts?
> 
> I could see a situation where a collaborator deletes a file. Or a user uploads a file and deletes it locally for some reason (to save space?). Maybe we can tell the users to not delete files locally until the report has been approved?

If we are gonna do that then a confirmation box would make the most sense to me. It doesn't help to warn them after they have already deleted it. Do we need some design for this?

File: frontend/src/components/FileUploader.js [hunk 1]
```diff
@@ -52,7 +52,7 @@ function Dropzone(props) {
       onChange(values);
     });
   };
-  const { getRootProps, getInputProps } = useDropzone({ onDrop });
+  const { getRootProps, getInputProps } = useDropzone({ onDrop, accept: 'image/*, .pdf, .docx, .xlsx, .pptx, .doc, .xls, .ppt, .zip' });
```
Comment (dcmcand) [inline frontend/src/components/FileUploader.js:55]: Adds file filtering on the frontend.

File: frontend/src/components/FileUploader.js [hunk 2]
```diff
@@ -78,7 +78,34 @@ Dropzone.propTypes = {
...
+      }>
+      <p>Are you sure you want to delete {file.originalFileName}?</p><p>This action cannot be undone.</p>
+    </Modal>
+  )
+    }
+
+  const ConnectedDeleteFileModal = connectModal(deleteFileModal)
+
+  return (
   <div className="files-table--container margin-top-2">
```
Comment (dcmcand) [inline frontend/src/components/FileUploader.js:109]: I got the basic layout for the code from https://trussworks.github.io/react-uswds/?path=/docs/other-modal--full-example and the basic language from https://preview.uxpin.com/ad995fd722bd26c0f9a1c8a115f1b41833637308#/pages/134880467/simulate/no-panels?mode=i

**Review by jasalisbury — APPROVED** [review]
Looks good!

**Review by rahearn — CHANGES_REQUESTED** [review]
I'll complete my review after we get the workers stuff merged into this, but in the meantime we'll need to add the authorization check on the delete handler.

**Review by rahearn — COMMENTED** [review]
This doesn't appear to be working for me. I hit the trash icon and it advances to the next form page and leaves the file where it is.

https://user-images.githubusercontent.com/285842/109431454-c16b6280-79d4-11eb-826f-40c9ca2965cf.mov

**Comment by dcmcand** [comment]
> This doesn't appear to be working for me. I hit the trash icon and it advances to the next form page and leaves the file where it is.
> Screen.Recording.2021-02-28.at.2.50.00.PM.mov

It is all working locally for me. Did you run `yarn docker:reset` and pull down all the latest changes? I know those are silly questions, but I have made that mistake before.

**Comment by dcmcand** [comment]
> This doesn't appear to be working for me. I hit the trash icon and it advances to the next form page and leaves the file where it is.
> Screen.Recording.2021-02-28.at.2.50.00.PM.mov

If you have some free time today, maybe we could pair to figure out what is going on there.

**Comment by rahearn** [comment]
> It is all working locally for me

I got it to work correctly for at least one path, but try this scenario:

1) Assign a grantee and save the report (you can, but don't have to, fill in the rest of the step 1 fields)
2) Assign at least one topic, upload a file, and click "Save & Continue"
3) Go back to the topics/resources page with the sidenav
4) Try to delete a file.

**Comment by dcmcand** [comment]
> > It is all working locally for me
> 
> I got it to work correctly for at least one path, but try this scenario:
> 
>     1. Assign a grantee and save the report (you can, but don't have to, fill in the rest of the step 1 fields)
> 
>     2. Assign at least one topic, upload a file, and click "Save & Continue"
> 
>     3. Go back to the topics/resources page with the sidenav
> 
>     4. Try to delete a file.

Ok, I see what you mean. I will check this out. Nice find.

**Review by rahearn — CHANGES_REQUESTED** [review]
Works and looks good, but there are 2 (related) accessibility concerns with the delete file modal.

1) When you use a keyboard to click the trash can, focus remains on that button, and you have to go backwards to activate the delete or cancel buttons
2) Screenreaders (or at least VoiceOver on macOS) do not announce that the modal has appeared, so a blind user will not have any indication that the delete button did anything.

I _think_ resetting focus to the header in the modal should fix both of those problems at once.

**Comment by dcmcand** [comment]
> Works and looks good, but there are 2 (related) accessibility concerns with the delete file modal.
> 
>     1. When you use a keyboard to click the trash can, focus remains on that button, and you have to go backwards to activate the delete or cancel buttons
> 
>     2. Screenreaders (or at least VoiceOver on macOS) do not announce that the modal has appeared, so a blind user will not have any indication that the delete button did anything.
> 
> 
> I _think_ resetting focus to the header in the modal should fix both of those problems at once.

Ok,  focus shifts to the modal on open now. From the reading I did, it seems like as long as it is inside the modal that should be fine, so I had it shift to the cancel button. I figured that was the safer option. However, I am far from an accessibility expert, so if this needs more work, please let me know.

**Review by rahearn — APPROVED** [review]
I think we might want to revisit this again. Ideally it would read out the entire dialog, but we can do that as a new issue.

**Pull request merged**